### PR TITLE
Add different Han fonts to EditorFonts for CJK

### PR DESCRIPTION
**Pull request by SaracenOne**
https://github.com/godotengine/godot-proposals/issues/3766

Okay, this PR is my attempt to address an issue which was raised by @TokageItLab in DMs about a localisation issue which occurs between Japanese, Korean, and both Traditional and Simplified Chinese. The topic is quite complex and is related to the way in which the Unicode consortium decided to assign matching codepoints for the vast amount of Han characters used in different languages, but which can be represented by different-looking glyphs depending on the language. I've included links which may help explain the issue.

https://en.wikipedia.org/wiki/Han_unification
https://heistak.github.io/your-code-displays-japanese-wrong/

On the web, the solution to this seems to be via using the 'lang' meta tag in conjunction with font sets for the respective languages. Since Godot now includes an equivilent tagging system for text and fonts, including different Noto font sets for CJK languages and tagging them appropriately seemed to be the most obvious solution.

However, I am leaving this PR in draft form for the time being for a few reasons:

Firstly, I am not a native speaker of any of the languages covered here, so there may be many issues which I may not be fully understanding. That being said, I invite those actively familiar with these languages to assess this issue and give feedback.

Secondly, I am concerned that the inclusion of several other large font sets into the binary could cause significant bloat to Godot's binary size, and I wonder if there be smarter ways to address this. Perhaps a step in the preprocess routine to strip duplicate han characters from the sets, or perhaps requiring that full support for these particular languages is obtained via external localisation packs.

### CODE REVIEW

File: editor/editor_fonts.cpp [hunk 1]
```diff
@@ -262,10 +268,54 @@ void editor_register_fonts(Ref<Theme> p_theme) {
...
+	FontChineseTraditional->set_language_support_override("ko", false);
+	FontChineseTraditionalBold->set_language_support_override("ko", false);
+	FontChineseTraditional->set_language_support_override("zh_CN", false);
+	FontChineseTraditionalBold->set_language_support_override("zh_CN", false);
+	FontChineseTraditional->set_language_support_override("zh_TW", true);
+	FontChineseTraditionalBold->set_language_support_override("zh_TW", true);
+
 	/* Droid Sans */
 
 	Ref<FontData> FontFallback = load_cached_internal_font(_font_DroidSansFallback, _font_DroidSansFallback_size, font_hinting, font_antialiased, true);
```
Comment (bruvzg) [inline editor/editor_fonts.cpp:318]: `DroidSansFallback` probably can be removed as well, it seems to be used for CJK characters only.

Comment (SaracenOne) [reply editor/editor_fonts.cpp:318]: I think so, yeah. I'll update the PR in a bit

Comment (akien-mga) [reply editor/editor_fonts.cpp:318]: You can then also remove the match section in https://github.com/godotengine/godot/blob/master/thirdparty/README.md#fonts

Comment (SaracenOne) [reply editor/editor_fonts.cpp]: Yeah, I'll remove those also. Actually, I see OpenSans is included too. Is that actually used anywhere?

Comment (bruvzg) [reply editor/editor_fonts.cpp]: > Is that actually used anywhere?

It's for export templates.

Comment (SaracenOne) [reply editor/editor_fonts.cpp]: I'll leave that one be then and just remove all the DroidSans stuff

**Review by timothyqiu — APPROVED** [review]
I've tried zh_CN, zh_TW, and jp locales, and the glyphs are rendered correctly.

But I'm not sure if this is worth increasing 45 MB for.

p.s. I also encountered the same issue as CI "Linux Builds / Editor with doubles and sanitizers" when building with multiple cores `-jN`.

**Comment by SaracenOne** [comment]
I'm totally on-board with the concern about bloating the binary size, hence why I'm hesitant to offer this PR as the primary solution to this problem. I do feel downloadable 'localisation pack' idea in the proposal would be better in the interest keeping the binary size nice and lean, but discussion will need to happen as to how to go about implementing that.

Not sure what's going on with the Linux build. Having trouble parsing the specific error, but I'll look into it a bit more later

**Comment by Zireael07** [comment]
Hm, can Godot maybe detect that editor or system language is one of CJK and suggest downloading optional fonts? This way, CJK speakers get improved looks, while all the others do NOT get unnecessary bloat...

**Comment by bruvzg** [comment]
> Not sure what's going on with the Linux build. Having trouble parsing the specific error, but I'll look into it a bit more later

A font header might be too big for GCC, I had a similar issue when I was trying to include a full ICU data file (~25 MB) as a header.

**Comment by SaracenOne** [comment]
> Hm, can Godot maybe detect that editor or system language is one of CJK and suggest downloading optional fonts? This way, CJK speakers get improved looks, while all the others do NOT get unnecessary bloat...

I think that would be a pretty good way of approaching it. I'm trying to think of other examples of software which does this, but I'm sure I've seen precedent for this approach.

> > Not sure what's going on with the Linux build. Having trouble parsing the specific error, but I'll look into it a bit more later
> 
> A font header might be too big for GCC, I had a similar issue when I was trying to include a full ICU data file (~25 MB) as a header.

Hmm, that sounds about right. I wonder if that's a sign that this is the wrong approach and indeed, the correct way is prompting the user to download the fonts instead.

**Comment by timothyqiu** [comment]
@SaracenOne I think you can try using one of the `NotoSansCJK<lang>-{Regular,Bold}.otf` fonts instead of the subset TTFs. Regular and Bold are about 32 MB in total (saves ~10 MB).

These fonts allow picking the appearance using `locl` OpenType feature. Different `<lang>` in the filename only decides the default appearance of glyphs.

**Comment by SaracenOne** [comment]
Cool, I'll look into this!

**Comment by bruvzg** [comment]
> These fonts allow picking the appearance using locl OpenType feature. Different <lang> in the filename only decides the default appearance of glyphs.

HarfBuzz have `locl` feature enabled by default, and it seems to be working fine with the fonts taken from https://github.com/googlefonts/noto-cjk/tree/main/Sans/OTF/Japanese. That's probably a better way to do it.

Screenshot 2022-01-07 at 09 41 41

> Hmm, that sounds about right. I wonder if that's a sign that this is the wrong approach and indeed, the correct way is prompting the user to download the fonts instead.

I guess if we go with downloading, it should be full "Language Pack" not just font, and will save even more space since editor/class db translations are quite large.

> A font header might be too big for GCC, I had a similar issue when I was trying to include a full ICU data file (~25 MB) as a header.

For the reference, adding 30 MB of fonts to test `locl` caused Apple clang to consume almost 6 GB of RAM, GCC is usually significantly more memory hungry, probably it's just running out of memory. I'm not sure how to solve it.

**Comment by reduz** [comment]
Can the newly added system fonts support improve this issue?

**Comment by bruvzg** [comment]
> Can the newly added system fonts support improve this issue?

If you have a list of pre-installed CJK fonts for each system. Or if you are willing to load every single system font and check if it supports all required characters (font metadata is not fully reliable, so getting supported language list might be not enough).

**Comment by bruvzg** [comment]
> Or if you are willing to load every single system font and check if it supports all required characters

For the reference, loading all system fonts and checking script support take less than a second, but results are quite unreliable, since it will include stuff like symbol and emoji fonts. So as character checks are definitely needed (still takes only a few seconds to get a full list).

Sample code:
```c++
String sample = U"一人大中的上出生不年自子地日本同下三小前所是我有了在国到会你他要以時也就可之得十事好那能学家多二和後用天者而心行新看文如道去都想方只手成問然当作主學这資長會来五這個个社市说们月为四為九交來政系業分时";
//from https://en.wiktionary.org/wiki/User:Hippietrail/Frequency:Most_Useful_CJK_Characters
Vector<String> fonts = OS::get_singleton()->get_system_fonts();
for (const String &E : fonts) {
	String path = OS::get_singleton()->get_system_font_path(E);
	if (!path.is_empty()) {
		Ref<FontFile> f;
		f.instantiate();
		if (f->load_dynamic_font(path) == OK) {
			if (f->is_script_supported("Hani")) {
				bool ok = true;
				for (int j = 0; j < sample.size(); j++) {
					bool has_char = f->has_char(sample[j]);
					ok = ok && has_char;
					if (!has_char) {
						break;
					}
				}
				if (ok) {
					print_line(vformat("Font: %s %s", E, path));
				}
			}
		}
	}
}
```

Result on macOS:

```
Font: Hannotate SC /System/Library/AssetsV2/com_apple_MobileAsset_Font7/8dc7805506cc9f233dcc19aabf593196842a47ae.asset/AssetData/Hannotate.ttc
Font: Hannotate TC /System/Library/AssetsV2/com_apple_MobileAsset_Font7/8dc7805506cc9f233dcc19aabf593196842a47ae.asset/AssetData/Hannotate.ttc
Font: HanziPen SC /System/Library/AssetsV2/com_apple_MobileAsset_Font7/9fdda46cbe802833590494a09b2787378340c597.asset/AssetData/Hanzipen.ttc
Font: HanziPen TC /System/Library/AssetsV2/com_apple_MobileAsset_Font7/9fdda46cbe802833590494a09b2787378340c597.asset/AssetData/Hanzipen.ttc
Font: Heiti SC /System/Library/Fonts/STHeiti Light.ttc
Font: Heiti TC /System/Library/Fonts/STHeiti Light.ttc
Font: Hiragino Sans GB /System/Library/Fonts/Hiragino Sans GB.ttc
Font: LingWai SC /System/Library/AssetsV2/com_apple_MobileAsset_Font7/e2c331b942cb338404a160a97a7cd6e8a31428e8.asset/AssetData/LingWaiSC-Medium.otf
Font: LingWai TC /System/Library/AssetsV2/com_apple_MobileAsset_Font7/b8cf53b3591d6062cd536ba5129454cba899a078.asset/AssetData/LingWaiTC-Medium.otf
Font: PingFang HK /System/Library/Fonts/PingFang.ttc
Font: PingFang SC /System/Library/Fonts/PingFang.ttc
Font: PingFang TC /System/Library/Fonts/PingFang.ttc
Font: SimSong /System/Library/AssetsV2/com_apple_MobileAsset_Font7/857d6c90171c328a4892c1492291d34e401d7f25.asset/AssetData/SimSong.ttc
Font: STFangsong /System/Library/AssetsV2/com_apple_MobileAsset_Font7/1821952872c81043711aab6910052b65da8edf2c.asset/AssetData/STFANGSO.ttf
Font: STHeiti /System/Library/AssetsV2/com_apple_MobileAsset_Font7/f0706a236683628e16427c6569e441423faaaa93.asset/AssetData/STXIHEI.ttf
Font: Wawati SC /System/Library/AssetsV2/com_apple_MobileAsset_Font7/18189590ed3a5f46cef20ed4d1cec2611dca13ff.asset/AssetData/WawaSC-Regular.otf
Font: Weibei SC /System/Library/AssetsV2/com_apple_MobileAsset_Font7/a8805b853edf89283dec908153a0015bdd1d00f8.asset/AssetData/WeibeiSC-Bold.otf
Font: Yuppy SC /System/Library/AssetsV2/com_apple_MobileAsset_Font7/96af7ec9e88d5dae450d3162213f92a7b1129430.asset/AssetData/YuppySC-Regular.otf
Font: Yuppy TC /System/Library/AssetsV2/com_apple_MobileAsset_Font7/e28431125242b9a298de3296370abdab4a7e8666.asset/AssetData/YuppyTC-Regular.otf
```

**Comment by akien-mga** [comment]
I believe the consensus here is that this is adding way too much to the binary size. If this is wanted and not already superseded by system fonts, we should look into this as part of making localization packages (https://github.com/godotengine/godot-proposals/issues/3421).

**Pull request closed without merging**